master/internal/db/postgres.go

```diff
@@ -23,12 +26,25 @@ type PgDB struct {
 	url       string
 }
 
+// Bun is our singleton database connection through the bun library. bun is the database library we
+// have decided to use for new code in the future due to its superior composability over bare SQL,
+// and its superior flexibiliy over e.g. gorm.  New code should not use the old bare SQL tooling.
+var Bun *bun.DB
+
 // ConnectPostgres connects to a Postgres database.
 func ConnectPostgres(url string) (*PgDB, error) {
 	numTries := 0
 	for {
 		sql, err := sqlx.Connect("pgx", url)
 		if err == nil {
+			Bun = bun.NewDB(sql.DB, pgdialect.New())
```

Comment:
@ioga a singleton that needs initialization is a bit stranger than one that does not. wydt looking at just this? should we at least wrap ConnectPostgres in a `sync.Once`?

Reply:
I am fine with this. the way the master code is structured, `ConnectPostgres` is called only once. adding an extra layer of stuff doesn't seem necessary to me: it doesn't solve anything, and the "singleton" comment already makes it pretty clear that users should use it as a singleton.

Reply:
I put it here because, as @ioga says, this is already called once, very early, by design.  So "it works".

I don't know how you would write a singleton that doesn't need initialization.  The master config has to be used to set this up.  I could make the singleton a function, like `db.Bun()`, that panics with an error if you haven't set up the singleton?

The only place I could see that happening is in unit tests, if e.g. you have some `TrialDB` interface you give to the trial actor, and some new code tries to use `db.Bun` directly.  Wdyt?

Reply:
> I don't know how you would write a singleton that doesn't need initialization.

if it does initialization in `init()` or doesn't need initialization.

i was just pointing out that a singleton that needs initialization at runtime is different than one that doesn't because it has the propensity to be misused whereas `net/http`'s `DefaultClient` does not. i understand it always _should_ be initialized once and early because of how the master code is structured but i think the difference between a singleton and explicitly injecting dependencies here is that a refactor of initialization code can now break at runtime whereas the other way it breaks at compile time. of course, if we had an issue where we refactored core.go so that we made db calls before the db was ready, i think we would catch it, but i'm evaluating singletons as a pattern, not just this specific instance.

concretely, i'm fine with singletons, but I think we at least wrap it in a `sync.Once` so it is at least impossible to accidentally reinitialize it, or even panic if `ConnectPostgres` is called twice, idk. with sync.once then it would be easier to reuse between tests without some strange fixture or reinitialization it every time, yes.

Reply:
I went with a `db.Bun()` function that panics if it is used wrong.  That should only happen in cases of gross developer error.

I agree with @ioga that we'd most likely never have problems if didn't write any protection around the singleton,  but even more than that, I like code that is bulletproof, and it's nearly zero overhead to do that here.